### PR TITLE
KIC 2.0 Integration Test Happy Path for v1.Ingress

### PR DESCRIPTION
**Pull request by shaneutt**
This PR adds the happy path integration tests and includes some newly completed [Kubernetes Testing Framework (KTF)](https://github.com/kong/kubernetes-testing-framework) functionality from upstream to provide a very basic Golang integration test for `v1.Ingress`.

This resolves https://github.com/Kong/kubernetes-ingress-controller/issues/1044.

- [x] supportive changes to [KTF](https://github.com/kong/kubernetes-testing-framework) and other utilities completed
- [x] testing "happy path" for `v1.Ingress` established
- [x] [KTF PR with new changes](https://github.com/Kong/kubernetes-testing-framework/pull/3) merged (blocker)
- [ ] integration tests plugged into `/.github/workflows` and actions created to exercise
- [ ] update the relevant [KEP](https://github.com/Kong/keps/blob/main/keps/0004-kong-kubernetes-testing.md) with the new context from this exercise

### CODE REVIEW

File: pkg/sendconfig/sendconfig.go
Comment (shaneutt) [inline pkg/sendconfig/sendconfig.go:182]: note: done due to deck `v1.5.0` update which was not backwards compat, but only required this change.

File: railgun/cmd/generators/controllers/networking/main.go
Comment (shaneutt) [inline railgun/cmd/generators/controllers/networking/main.go:221]: note: I renamed this because we were only ever dealing with `Ingress` APIs with this function.

File: railgun/controllers/configuration/ingress_utils.go
Comment (shaneutt) [inline railgun/controllers/configuration/ingress_utils.go:21]: note: I made this public so that `main.go` could use it to short circuit the `UDPIngress` API on clusters which did not support it.

File: railgun/controllers/configuration/secret_utils.go
Comment (shaneutt) [inline railgun/controllers/configuration/secret_utils.go:31]:
```suggestion
	// TODO need EVENTS here
	// TODO need more status updates
	// TODO: (shane) I want to refactor this into several smaller functions
	// ^ follow up for these items is in: https://github.com/Kong/kubernetes-ingress-controller/issues/1094
```

File: railgun/controllers/configuration/secret_utils.go
Comment (shaneutt) [inline railgun/controllers/configuration/secret_utils.go:104]:
```suggestion
// storeRuntimeObject stores a runtime.Object in the configuration secret. Callers should re-queue after this completes successfully.
```

File: railgun/controllers/configuration/secret_utils.go
Comment (shaneutt) [inline railgun/controllers/configuration/secret_utils.go:122]:
```suggestion
	// TODO need more status updates
	// TODO: (shane) I want to refactor this into several smaller functions
	// ^ follow up for these items is in: https://github.com/Kong/kubernetes-ingress-controller/issues/1094
```

File: railgun/test/integration/ingress_controllers_test.go
Comment (shaneutt) [inline railgun/test/integration/ingress_controllers_test.go:26]: note: while most of the work in this PR was supportive, this is the "happy path" that the supportive work was attempting to accomplish. The goal is that going forward by separating the environment into [it's own concerns](https://github.com/kong/kubernetes-testing-framework) this would be all that is needed for a contributor to add an integration test for the API they worked on.

File: .github/workflows/integration_test.yaml
Comment (shaneutt) [inline .github/workflows/integration_test.yaml:22]: Eventually we'll configure our public images for anonymous access, until then we need to authenticate with the Packages Registry.

**Comment by rainest** [comment]
My approach to reviewing this was to try and add a new test to understand the workflow and see what were problems or points of interest for me. I failed! But I learned things.

**Separation of concerns between runbooks and tests**

IMO this is sort of blocking unless I'm missing something. Adding new tests would be difficult if they follow the proposed pattern, which bakes a lot of logic into KTF.

The ingress runbook from https://github.com/Kong/kubernetes-testing-framework/pull/3 has configuration that we'd want to set per test. Existing integration tests define at least an Ingress resource and usually something else, though I'm ignoring the something else for now since we don't have our CRDs in the framework (AFAIK that's expected for this PR, and they'll come later).

`runbooks.DeployIngressForContainer()` defines the Ingress for you though, so the test is limited to what that runbook provides. If it doesn't (let's say you want to add a test for the `konghq.com/https-redirect-status-code` annotation), you need a new runbook, and would be blocked on a PR to a repo other than KIC, and then more PRs to the other repo if you fail to get your runbook correct the first go round.

Is there a reason we can't/shouldn't call the framework functions from tests directly? My [attempt to add a redirect annotation integration test](https://gist.github.com/rainest/0a77e7e8244d4e9a2fc15cb1f5a9cda9#file-redirect_test-go) seems fine, albeit a bit more verbose. More on that later (it didn't actually work), but in general this seems like the level of control we want over the tests in this repo.

We definitely need a separate KTF PR here, since it didn't have the generators before, but for additional tests I'd expect that you wouldn't normally need a KTF PR unless you're adding new generator functionality (e.g. we'll probably need the Service generator to accept an annotation map also) or something low level (e.g. the proposed cluster runbook isn't something I'd expect to change between tests).

**Tests don't clean up after themselves**

Resources remain deployed after their tests complete, and subsequent tests that attempt to use similar configuration will fail. For example, duplicating `TestIngress` into an addtional `TestIngressAgain` results in a failure because the deployment is already present:

```
=== RUN   TestIngress
--- PASS: TestIngress (54.04s)
=== RUN   TestIngressAgain
    ingress_controllers_test.go:35: 
        	Error Trace:	ingress_controllers_test.go:35
        	Error:      	Received unexpected error:
        	            	deployments.apps "nginx" already exists
        	Test:       	TestIngressAgain
--- FAIL: TestIngressAgain (1.07s)
FAIL
```

You can probably invoke `kc.NetworkingV1().Ingresses("default").Delete(ctx, ingress, opts)` and such on everything manually, but ideally you can quickly clean up everything in one step. Is there a good ready-made way to track resources for individual tests and nix them after completion?

If not, should we go with something imperfect but quick, e.g. create a namespace for each test and delete it after? That can probably handle most of what we need, though not cluster-wide resources. I think that'd only be a problem for KongClusterPlugin.

**How do we debug tests?**

This was one of my main points of frustration with the existing integration test system, and is much better here.

My initial attempt to add a new test was unsuccessful, though the reason was unclear. Basic usage of `assert.Eventually()` won't shed much light on this, since it will just sit around for a while and then fail if its condition wasn't met. In the test above, I'll see a failure if `resp.StatusCode == http.StatusMovedPermanently` never ends up evaluating to `true`, but won't get any indication of what the status was otherwise.

Go tests do at least provide a much easier debug route than bash, since I can now use Delve on tests that don't work, though it's somewhat limited because the timeout still applies and you'll eventually get kicked out if you don't look around long enough. We can presumably add logging to tests in the future for lighter-touch debugging, though I didn't try that approach immediately.

Local non-code debugging worked well. It's easy to check the cluster services, get the MetalLB address, and send requests to that.

**Should separate timeouts for initial Kong deployment spawn and test cases**

QoL thing that we should keep in mind, but don't need to address now. At 5m, `IngressTimeout` leaves ample time for Kong to come online (on my machine, the first test took about a minute to complete), but is rather long for tests after. If your tests don't succeed, you'll need to wait a while for the failure to finalize, even though realistically the test client will just repeat a request that doesn't work.

**New test didn't work. Wat**

This is just... odd. My attempt at creating a new test seems fine at a glance, but I couldn't make it succeed. I'm not sure why. Go HTTP is [reporting that it couldn't reach the host requested](https://gist.github.com/rainest/0a77e7e8244d4e9a2fc15cb1f5a9cda9#file-unreachable-txt), which seems not in line with what I can see elsewhere. tcpdump, [shows that the test client's request was successful and did get the expected response](https://gist.github.com/rainest/0a77e7e8244d4e9a2fc15cb1f5a9cda9#file-301_pcap-txt), so I am quite confused.

This issue doesn't happen if I add a duplicate of `TestIngress` with a different Deployment name, though I'm unsure why wrapping resource creation in a function would have made a difference: ultimately that half of the test just creates resources (which definitely works even for the broken test), and the Go HTTP client doesn't have any knowledge of exactly how the resources were created--it's just sending a request to an arbitrary URL.

Not the same run above, but [wat.tar.gz](https://github.com/Kong/kubernetes-ingress-controller/files/6153040/wat.tar.gz) contains a dump of both tests, and they don't look different at the network level: both show packets over both br-73cea2bdf38c and veth8874f26, all of which are ACKed as expected.

**Comment by shaneutt** [comment]
@rainest thank you for your very thorough and detailed review :bow: 

> Is there a reason we can't/shouldn't call the framework functions from tests directly?

There's no explicit reason, we can make adjustments as we see fit.

> My [attempt to add a redirect annotation integration test](https://gist.github.com/rainest/0a77e7e8244d4e9a2fc15cb1f5a9cda9#file-redirect_test-go) seems fine, albeit a bit more verbose. More on that later (it didn't actually work), but in general this seems like the level of control we want over the tests in this repo.

I can be agreeable to this. The runbooks are meant to cover really common cases but this level of control seems legitimate as well.

**Action:** I will plug your test in to this PR and get it functional.

> Resources remain deployed after their tests complete, and subsequent tests that attempt to use similar configuration will fail.
> You can probably invoke `kc.NetworkingV1().Ingresses("default").Delete(ctx, ingress, opts)` and such on everything manually, but ideally you can quickly clean up everything in one step. Is there a good ready-made way to track resources for individual tests and nix them after completion?
> If not, should we go with something imperfect but quick, e.g. create a namespace for each test and delete it after? That can probably handle most of what we need, though not cluster-wide resources. I think that'd only be a problem for KongClusterPlugin.

This is a known deficiency of the testing environment as it already existed: tests have to clean up after themselves.

The following issue has been created as the **follow up** for this:

https://github.com/Kong/kubernetes-testing-framework/issues/4

> Go tests do at least provide a much easier debug route than bash, since I can now use Delve on tests that don't work, though it's somewhat limited because the timeout still applies and you'll eventually get kicked out if you don't look around long enough.

To help with this, I have created the a **follow up** issue which defines making timeouts configurable and otherwise allowing the testing environment to influence them as per its own needs:

https://github.com/Kong/kubernetes-testing-framework/issues/5

> We can presumably add logging to tests in the future for lighter-touch debugging, though I didn't try that approach immediately.

Another reasonable **follow up** item:

https://github.com/Kong/kubernetes-testing-framework/issues/6

> Local non-code debugging worked well. It's easy to check the cluster services, get the MetalLB address, and send requests to that.

:+1: glad to hear it!

> QoL thing that we should keep in mind, but don't need to address now. At 5m, `IngressTimeout` leaves ample time for Kong to come online (on my machine, the first test took about a minute to complete), but is rather long for tests after. If your tests don't succeed, you'll need to wait a while for the failure to finalize, even though realistically the test client will just repeat a request that doesn't work.

It is my hope that the previously mentioned improvements to timeouts should be a reasonable **follow-up** for this:

https://github.com/Kong/kubernetes-testing-framework/issues/5

> This is just... odd. My attempt at creating a new test seems fine at a glance, but I couldn't make it succeed. I'm not sure why. Go HTTP is [reporting that it couldn't reach the host requested](https://gist.github.com/rainest/0a77e7e8244d4e9a2fc15cb1f5a9cda9#file-unreachable-txt), which seems not in line with what I can see elsewhere. tcpdump, [shows that the test client's request was successful and did get the expected response](https://gist.github.com/rainest/0a77e7e8244d4e9a2fc15cb1f5a9cda9#file-301_pcap-txt), so I am quite confused.
> This issue doesn't happen if I add a duplicate of `TestIngress` with a different Deployment name, though I'm unsure why wrapping resource creation in a function would have made a difference: ultimately that half of the test just creates resources (which definitely works even for the broken test), and the Go HTTP client doesn't have any knowledge of exactly how the resources were created--it's just sending a request to an arbitrary URL.
> Not the same run above, but [wat.tar.gz](https://github.com/Kong/kubernetes-ingress-controller/files/6153040/wat.tar.gz) contains a dump of both tests, and they don't look different at the network level: both show packets over both br-73cea2bdf38c and veth8874f26, all of which are ACKed as expected.

**Action:** I'll help dig into this, thank you for being so thorough and bringing this up :vulcan_salute:

**Comment by shaneutt** [comment]
@rainest the above commits remove the "runbooks" concept in favor of a modular "cluster" interface, let me know if the new way the tests are written seem preferable.

**Comment by shaneutt** [comment]
@rainest I added your test here: https://github.com/Kong/kubernetes-ingress-controller/pull/1102/commits/ac86ece760993558f4f81cef17361e6da1be7ed9

The reason it was failing is that the proxy on the cluster has HTTPS disabled (explicitly), so the packets could never make it to the proxy server itself.

To keep scope low how would you feel about us implementing another test, and then we can add HTTPS support for tests in as a **follow-up**?

File: railgun/controllers/configuration/secret_utils.go
Comment (mflendrich) [inline railgun/controllers/configuration/secret_utils.go:34]:
```suggestion
	// if this is an Ingress resource make sure it's managed by KIC
```

File: railgun/controllers/configuration/secret_utils.go
Comment (mflendrich) [inline railgun/controllers/configuration/secret_utils.go:42]: It doesn't feel right to me that this function operates on namespaced names in two completely different fashions:

ingress namespaced name is a `types.NamespacedName`, while the secret namespaced name is constructed from `os.Getenv` and a const stored elsewhere.

Would it sound appropriate for this function to take two `types.NamespacedName` arguments? One for the (input) ingress being reconciled, and the other for the secret being updated?

File: railgun/controllers/configuration/secret_utils.go
Comment (mflendrich) [inline railgun/controllers/configuration/secret_utils.go:58]: Why is `Requeue: true` necessary here?

File: railgun/controllers/configuration/secret_utils.go
Comment (mflendrich) [inline railgun/controllers/configuration/secret_utils.go:67]: Tell me if I'm missing something, but if we `return` from this function after the finalizer is set, `storeRuntimeObject` further down this function will not be run, and thus the reconciliation event will get lost.

File: railgun/controllers/configuration/secret_utils.go
Comment (mflendrich) [inline railgun/controllers/configuration/secret_utils.go:102]:
```suggestion
	foundCFG, ok := secret.Data[key]
	return ok && bytes.Equal(foundCFG, cfg)
```

File: railgun/controllers/configuration/secret_utils.go
Comment (mflendrich) [inline railgun/controllers/configuration/secret_utils.go:117]: Are you confident that `Update` is concurrency-safe? Not saying it isn't, just double-checking.

File: railgun/controllers/configuration/secret_utils.go
Comment (mflendrich) [inline railgun/controllers/configuration/secret_utils.go:142]: I'd opt for making a migration from `Update` to `Patch` a blocker for the 2.0 release.

File: railgun/pkg/constants/constants.go
Comment (mflendrich) [inline railgun/pkg/constants/constants.go:6]: This is a CLI flag in KIC 1.x and needs to remain as such.

File: railgun/controllers/configuration/secret_utils.go
Comment (shaneutt) [reply railgun/controllers/configuration/secret_utils.go:42]: I may need help to understand the problem better: the secret namespace is explicitly provided via ENV right now, you would prefer this be a function argument is that right?

File: railgun/controllers/configuration/secret_utils.go
Comment (shaneutt) [reply railgun/controllers/configuration/secret_utils.go:58]: Kubernetes controller best practices: Generally try to only make one write to the API per reconcilation. This function is re-entrant and so the next queue of it will reach the next logic gate.

File: railgun/controllers/configuration/secret_utils.go
Comment (shaneutt) [reply railgun/controllers/configuration/secret_utils.go:67]: `obj.SetFinalizers` only changes the finalizers in local memory, the `c.Update` is the call actually responsible for updating the object in the Kubernetes API. Thus if we return here due to failure reconcilation will report the error and then requeue (throwing an error causes a requeue), so the logic here is thus:

```
   ┌────────────────┐
   │ GET Finalizers │◄──────┐
   └────────┬───────┘       │
            │               │ NO (ERROR)
┌───────────▼─────────────┐ │
│ Kong Finalizer Present? ├─┘
└───────────┬─────────────┘
            │ YES
        ┌───▼───┐
        │ Done! │
        └───────┘
```

File: railgun/controllers/configuration/secret_utils.go
Comment (shaneutt) [reply railgun/controllers/configuration/secret_utils.go:117]: Yes :+1:

File: railgun/controllers/configuration/secret_utils.go
Comment (shaneutt) [reply railgun/controllers/configuration/secret_utils.go:142]: Sounds good, https://github.com/Kong/kubernetes-ingress-controller/issues/1104 was created for this purpose :+1:

File: railgun/pkg/constants/constants.go
Comment (shaneutt) [reply railgun/pkg/constants/constants.go:6]: Sounds good, context added in https://github.com/Kong/kubernetes-ingress-controller/issues/1097

File: railgun/controllers/configuration/secret_utils.go
Comment (mflendrich) [reply railgun/controllers/configuration/secret_utils.go:42]: Yes, that summarizes my motivation well.

This preference is based on two assumptions:
- It would allow us to separate the use of the configuration parameter (secret namespace+name) from its presentation (env var + a code constant)
- It would unify presentation of two namespaced names used in this function, which is desirable.

File: railgun/controllers/configuration/secret_utils.go
Comment (mflendrich) [reply railgun/controllers/configuration/secret_utils.go:67]: Ah, I didn't notice that the `return` happens in the error case. I misread it (as if it `return`ed on success). My bad.

**Comment by shaneutt** [comment]
NOTE: CI wont pass for now due to lack of HTTPS listen in the container image's proxy, waiting on @rainest for some comment feedback and we'll get that taken care of later. Feel free to continue reviewing even though CI is red :vulcan_salute:

File: railgun/controllers/configuration/secret_utils.go
Comment (shaneutt) [reply railgun/controllers/configuration/secret_utils.go:42]: Sounds good. I've added this as one of the criteria for the follow up which covers CLI options conformity between v1 and v2:

https://github.com/Kong/kubernetes-ingress-controller/issues/1097

This conversation has been linked therein for context.

Let me know if that seems adequate to move forward or if you would like this handled differently.

Comment (mflendrich) [reply railgun/controllers/configuration/secret_utils.go:42]: Let's talk about this in our zoom tomorrow.

File: railgun/test/integration/ingress_controllers_test.go
Comment (rainest) [inline railgun/test/integration/ingress_controllers_test.go:84]: > The reason it was failing is that the proxy on the cluster has HTTPS disabled (explicitly), so the packets could never make it to the proxy server itself.

Ah, okay, TIL that the Go stdlib HTTP client is not like curl and handles redirects by default, which isn't what I want here: we care about the initial response, not wherever the request would end up. This _should_ work:

```suggestion

	client := &http.Client{
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	assert.Eventually(t, func() bool {
		u := proxyURL()
		resp, err := client.Get(fmt.Sprintf("%s/example", u.String()))
```

I'm now running into some other issue where the Kong deployment is no longer created at all, and am not sure what's causing that. It's baked into the KIND image, correct?

Wasn't sure if that was maybe tied to the KTF changes that happened this PR started (go.mod was still pointing to a version from the 17th and there'd been further changes today). I did a replace (`replace github.com/Kong/kubernetes-testing-framework => github.com/shaneutt/kubernetes-testing-framework v0.0.0-20210318144855-8506060a256f`) and attempted another run, which did spawn the Kong deployment, but failed due to some other issue. Subsequent runs stopped spawning the Kong deployment again.

**Comment by rainest** [comment]
Alright, following more local testing, I'm reasonably confident in the fix portion of [my comment on the redirect test](https://github.com/Kong/kubernetes-ingress-controller/pull/1102#discussion_r597285334), but am quite confused as to why everything else appears to have broken for me following the other changes, but works fine in CI.

The Kong deployment now frequently fails to spawn at all. The KIND container indicates that the job failed, though logs there don't provide a whole lot of interesting info:

```
# systemctl status kong-setup.service
● kong-setup.service - Setup Kong Kubernetes Ingress Controller
     Loaded: loaded (/etc/systemd/system/kong-setup.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Thu 2021-03-18 23:11:56 UTC; 51s ago
   Main PID: 695 (code=exited, status=1/FAILURE)

Mar 18 23:11:55 fdf38f2e-29bd-467e-a706-0d3546e78f98-control-plane systemd[1]: kong-setup.service: Control process exited, code=exited, status=1/
FAILURE
Mar 18 23:11:55 fdf38f2e-29bd-467e-a706-0d3546e78f98-control-plane systemd[1]: kong-setup.service: Failed with result 'exit-code'.
Mar 18 23:11:55 fdf38f2e-29bd-467e-a706-0d3546e78f98-control-plane systemd[1]: Failed to start Setup Kong Kubernetes Ingress Controller.
Mar 18 23:11:56 fdf38f2e-29bd-467e-a706-0d3546e78f98-control-plane systemd[1]: kong-setup.service: Scheduled restart job, restart counter is at 7.
Mar 18 23:11:56 fdf38f2e-29bd-467e-a706-0d3546e78f98-control-plane systemd[1]: Stopped Setup Kong Kubernetes Ingress Controller.
Mar 18 23:11:56 fdf38f2e-29bd-467e-a706-0d3546e78f98-control-plane systemd[1]: kong-setup.service: Start request repeated too quickly
.
Mar 18 23:11:56 fdf38f2e-29bd-467e-a706-0d3546e78f98-control-plane systemd[1]: kong-setup.service: Failed with result 'exit-code'.
Mar 18 23:11:56 fdf38f2e-29bd-467e-a706-0d3546e78f98-control-plane systemd[1]: Failed to start Setup Kong Kubernetes Ingress Controller.
```

Shelling in and starting it manually succeeds, though the local controller instance doesn't talk to it, apparently. Not sure how to see more info about that since [it no longer logs to a temp file](https://github.com/Kong/kubernetes-ingress-controller/commit/7832159764b146a9c611b690205738ec4d3f5add#diff-14dec2cd8ec27122d6d7d515bb4e159b8b8d07fba631a7379d355183a2997a98L87-L104). Maybe that wasn't working before? Since I hadn't encountered issues previously I hadn't tried to check.

Simply reverting to bdde23e doesn't appear to fix that, however, so something else appears to have gone haywire on my machine.

**Comment by shaneutt** [comment]
> Based on some un-scientific log watching, Systemd apparently tries to install Kong before the control plane is even online, which naturally fails, and it gives up after retrying too many times, so Kong never starts without manual intervention. What _should_ have ensured that it waits until K8S is running, or at least kept retrying until it succeeded? Not sure why that was working for me fine the other day, since it still seems like that should have been entirely up to the KIND image, which didn't change.

Sounds like we have some race conditions and potential fragile points in our provisioning logic that need to be sorted out :thinking: 

I haven't run into anything like this on my end but the current systemd unit approach to pre-deploying the proxy is a hack until that can be baked into the image proper, so I think naturally that's probably where changes need to occur.

I'll dig into this a little further today and do some testing to see if I can make the Kong deployment mechanism more resilient.

File: railgun/controllers/configuration/secret_utils.go
Comment (shaneutt) [inline railgun/controllers/configuration/secret_utils.go:31]:
```suggestion
	// TODO: (shane) I want to refactor this into several smaller functions
	// TODO: collapse nsn + obj, this is redudant as obj includes nsn
	// TODO: pass in secret namespace as part of function sig instead of env
```

File: railgun/controllers/configuration/secret_utils.go
Comment (shaneutt) [reply railgun/controllers/configuration/secret_utils.go:42]: follow up in: https://github.com/Kong/kubernetes-ingress-controller/pull/1102/files#r597713552

**Comment by shaneutt** [comment]
After some investigation @rainest I have reason to believe our problems may relate to issues with image pulling that we've experienced elsewhere recently: in its current state the images are pulled on _every_ test deploy (something that was intended to be taken care of in a later iteration of `kong/kind-images`) and I'm seeing situations where it appears to intermittently fail to pull on my systems.

I'll continue digging in, it may be that we need to go ahead and do the intended follow up now to improve this.

File: railgun/controllers/configuration/secret_utils.go
Comment (mflendrich) [inline railgun/controllers/configuration/secret_utils.go:37]:
```suggestion
	// BUG: this takes only the kind into account, not the API group.
	if obj.GetObjectKind().GroupVersionKind().Kind == "Ingress" {
```

**Review by mflendrich — APPROVED** [review]
Conditional LGTM, assuming that #1114 happens before 2.0 GA

**Comment by shaneutt** [comment]
https://github.com/kubernetes-sigs/kind/issues/2143 is somewhat relevant to improving our kind images for integration tests.

**Pull request merged**
